packages/kit/src/runtime/client/client.js

```diff
@@ -0,0 +1,1233 @@
+
+	let ready = false;
+	stores.session.subscribe(async (value) => {
+		$session = value;
+
+		if (!ready) return;
+		session_id += 1;
+
+		const intent = get_navigation_intent(new URL(location.href));
+		update(intent, [], true);
+	});
+	ready = true;
+
+	/** Keeps tracks of multiple navigations caused by redirects during rendering */
+	let navigating = 0;
+
+	let router_enabled = true;
+
+	// keeping track of the history index in order to prevent popstate navigation events if needed
+	let current_history_index = history.state?.[INDEX_KEY] ?? 0;
+
+	if (current_history_index === 0) {
+		// create initial history entry, so we can return here
+		history.replaceState({ ...history.state, [INDEX_KEY]: 0 }, '', location.href);
+	}
+
+	// if we reload the page, or Cmd-Shift-T back to it,
+	// recover scroll position
+	const scroll = scroll_positions[current_history_index];
+	if (scroll) scrollTo(scroll.x, scroll.y);
+
+	let hash_navigating = false;
+
+	/** @type {import('types').Page} */
+	let page;
+
+	/** @type {{}} */
+	let token;
+
+	/** @type {{}} */
+	let navigating_token;
+
+	/**
+	 * @param {string} href
+	 * @param {{ noscroll?: boolean; replaceState?: boolean; keepfocus?: boolean; state?: any }} opts
+	 * @param {string[]} redirect_chain
+	 */
+	async function goto(
+		href,
+		{ noscroll = false, replaceState = false, keepfocus = false, state = {} },
+		redirect_chain
+	) {
+		const url = new URL(href, get_base_uri(document));
+
+		if (router_enabled) {
+			return navigate({
+				url,
+				scroll: noscroll ? scroll_state() : null,
+				keepfocus,
+				redirect_chain,
+				details: {
+					state,
+					replaceState
+				},
+				accepted: () => {},
+				blocked: () => {}
+			});
+		}
+
+		await traditional_navigation(url);
+	}
+
+	/** @param {URL} url */
+	async function prefetch(url) {
+		if (!owns(url)) {
+			throw new Error('Attempted to prefetch a URL that does not belong to this app');
+		}
+
+		const intent = get_navigation_intent(url);
+
+		loading.promise = get_navigation_result(intent, false);
+		loading.id = intent.id;
+
+		return loading.promise;
+	}
+
+	/**
+	 * @param {import('./types').NavigationIntent} intent
+	 * @param {string[]} redirect_chain
+	 * @param {boolean} no_cache
+	 * @param {{hash?: string, scroll: { x: number, y: number } | null, keepfocus: boolean}} [opts]
+	 */
+	async function update(intent, redirect_chain, no_cache, opts) {
+		const current_token = (token = {});
+		let navigation_result = await get_navigation_result(intent, no_cache);
+
+		if (!navigation_result && intent.url.pathname === location.pathname) {
+			navigation_result = await load_root_error_page({
+				status: 404,
+				error: new Error(`Not found: ${intent.url.pathname}`),
+				url: intent.url
+			});
+		}
+
+		if (!navigation_result) {
+			await traditional_navigation(intent.url);
+			return; // unnecessary, but TypeScript prefers it this way
+		}
+
+		// abort if user navigated during update
+		if (token !== current_token) return;
```

Comment:
I noticed when working on https://github.com/sveltejs/kit/pull/4101 that this token stuff looks broken in `master`. I didn't attempt to fix it and it looks like it's not addressed here either. Assuming I haven't missed something, `current_token` is only ever set to `{}` and so this is a no-op check

Reply:
`update()` can be called by a different navigation/redirect between the time it started and the time this check is made, meaning `token` will point to a different `{}` than this invocation's `current_token`, and in that case we abort (because this call to `update()` was superseded).

Reply:
it's working as intended as far as I can see. consider this simplified version:

```js
let token;

async function log(message) {
  const current_token = (token = {});
  await Promise.resolve();
  if (token !== current_token) return;

  console.log(message);
}

// only b is logged; a is aborted
log('a');
log('b');
```

Reply:
Oh. I see. That's a bit of a subtle thing that relies on `===` checking that they're the same instance. I was just looking at it and seeing that they were both always `{}` and doing more of a checking that the values were equivalent in my head. Maybe we could either add a comment or rename to `current_token_instance` and `token_instance`?

Reply:
I think it's clear enough, given the comment immediately above the check